docs/advanced.md

```diff
@@ -1,19 +1,257 @@
+{
+    public function process(Record $record, MiddlewareHandler $handler): void
+    {
+        // Here you can manipulate your record before handle it in your consumer
+
+        $handler->handle($record);
+    }
+}
+
+```
+
+You may overwrite the record payload by calling `$record->setPayload()`:
+
+```php
+public function process(Record $record, MiddlewareHandler $handler): void
+{
+    $payload = $record->getPayload();
+
+    $record->setPayload(json_decode($payload));
+
+    $handler->handle($record);
+}
+```
+
+Then you may configure this new middleware to be executed for every record by adding it on the config file `config/kafka.php`:
+
+```php
+// ...
+'middlewares' => [
+    'consumer' => [
+        \Metamorphosis\Middlewares\Log::class,
+        \App\Kafka\Middlewares\JsonDeserializer::class,
+    ],
+],
+// ...
+```
+
+If you wish, you may set a middleware to run of a topic level or a consumer group level:
+
+```php
+'topics' => [
+    'price-update' => [
+        'topic' => 'products.price.update',
+        'broker' => 'price-brokers',
+        'consumer-groups' => [
+            'default' => [
+                'offset' => 'initial',
+                'consumer' => '\App\Kafka\Consumers\PriceUpdateConsumer',
+                'middlewares' => [
+                    \App\Kafka\Middlewares\ConsumerGroupMiddlewareExample::class,
+                ],
+            ],
+        ],
+        'middlewares' => [
+            \App\Kafka\Middlewares\TopicMiddlewareExample::class,
+        ],
+    ],
+],
+```
+
+The order matters here, they'll be execute as queue, from the most global scope to the most specific (global scope > topic scope > group-consumers scope).
+
+
+<a name="commands"></a>
+### Commands
+There's a few commands to help automate the creation of classes and to run the consumer.
+
+<a name="commands-consumer"></a>
+#### Creating Consumer
+You can create a consumer class, that will handle all records received from the topic using the follow command:
+```bash
+$ php artisan make:kafka-consumer PriceUpdateConsumer
+```
+This will create a KafkaConsumer class inside the application, on the `app/Kafka/Consumers/` directory
+
+There, you'll have a `handler` method, which will send all records from the topic to the Consumer.
+Methods will be available for handling exceptions:
+ - `warning` method will be call whenever something not critical is received from the topic.
+    Like a message informing that there's no more records to consume.
+ - `failure` method will be call whenever something critical happens, like an error to decode the record.
+
+```php
+use App\Repository;
+use Exception;
+use Metamorphosis\Exceptions\ResponseWarningException;
+use Metamorphosis\Record;
+use Metamorphosis\TopicHandler\Consumer\AbstractHandler;
+
+class PriceUpdateConsumer extends AbstractHandler
+{
+    public $repository;
+
+    /**
+     * Create a new consumer topic handler instance.
+     *
+     * @return void
```

Comment:
Constructors, destructors, etc, does not have the concept of return, even return `void` :)